src/coreclr/jit/importer.cpp

```diff
@@ -509,61 +507,70 @@ void Compiler::impAppendStmt(Statement* stmt, unsigned chkLevel, bool checkConsu
+        // can modify them and spill the references. In doing so, we make two assumptions:
+        //
+        // 1. All locals which can be modified indirectly are marked as address-exposed or with
+        //    "lvHasLdAddrOp" -- we will rely on "impSpillSideEffects(spillGlobEffects: true)"
+        //    below to spill them.
+        // 2. Trees that assign to unaliased locals are always top-level (this avoids having to
+        //    walk down the tree here), and are a subset of what is recognized here.
+        //
+        // If any of the above are violated (say for some temps), the relevant code must spill
+        // things manually.
+        //
+        LclVarDsc* dstVarDsc = nullptr;
+        if (expr->OperIs(GT_ASG) && expr->AsOp()->gtOp1->OperIsLocal())
         {
-            GenTreeFlags op2Flags = expr->AsOp()->gtOp2->gtFlags & GTF_GLOB_EFFECT;
-            assert(flags == (op2Flags | GTF_ASG));
-            flags = op2Flags;
+            dstVarDsc = lvaGetDesc(expr->AsOp()->gtOp1->AsLclVarCommon());
         }
-
-        if (flags != 0)
+        else if (expr->OperIs(GT_CALL, GT_RET_EXPR)) // The special case of calls with return buffers.
         {
-            bool spillGlobEffects = false;
+            GenTree* call = expr->OperIs(GT_RET_EXPR) ? expr->AsRetExpr()->gtInlineCandidate : expr;
 
-            if ((flags & GTF_CALL) != 0)
+            if (call->TypeIs(TYP_VOID) && call->AsCall()->TreatAsShouldHaveRetBufArg(this))
             {
-                // If there is a call, we have to spill global refs
-                spillGlobEffects = true;
-            }
-            else if (!expr->OperIs(GT_ASG))
-            {
-                if ((flags & GTF_ASG) != 0)
+                GenTree* retBuf;
+                if (call->AsCall()->ShouldHaveRetBufArg())
                 {
-                    // The expression is not an assignment node but it has an assignment side effect, it
-                    // must be an atomic op, HW intrinsic or some other kind of node that stores to memory.
-                    // Since we don't know what it assigns to, we need to spill global refs.
-                    spillGlobEffects = true;
+                    retBuf = call->AsCall()->gtArgs.GetRetBufferArg()->GetNode();
                 }
-            }
-            else
-            {
-                GenTree* lhs = expr->gtGetOp1();
-                GenTree* rhs = expr->gtGetOp2();
-
-                if (((rhs->gtFlags | lhs->gtFlags) & GTF_ASG) != 0)
+                else
                 {
-                    // Either side of the assignment node has an assignment side effect.
-                    // Since we don't know what it assigns to, we need to spill global refs.
-                    spillGlobEffects = true;
+                    assert(!call->AsCall()->gtArgs.HasThisPointer());
+                    retBuf = call->AsCall()->gtArgs.GetArgByIndex(0)->GetNode();
```

Comment:
This handling is a bit strange to me given the comments above since `TreatAsShouldHaveRetBufArg()` calls that are not `ShouldHaveRetBufArg()` are just normal calls that get an extra (completely standard, ABI-wise) argument that point to a local. Why is the special handling necessary for them? Where does it get handled instead for a call taking such a pointer explicitly?

Reply:
> Why is the special handling necessary for them?

This code mirrors what we have in `impStructAssignPtr` (in fact, it must be exactly in sync with it).

The basic premise is that both "fake" and "real" return buffer calls follow the same scheme where they are first manifested as actual IR values (`TYP_STRUCT`s). I. e. if could call these helpers in IL:
```
call <helper>
stloc lcl // 'lcl' will not be marked as "aliased"
```
Another way to look at it is that these (and only these) calls gain an additional side effect of assigning to a possibly "unaliased" local. This is analogous to how we must mark `GTF_CALL_M_RETBUFFARG_LCLOPT` with `GTF_ASG`.

> Where does it get handled instead for a call taking such a pointer explicitly?

For `Call(&a)`, `a` will be marked as "aliased", or spilled explicitly.

Reply:
Got it, that makes sense.

It may be good to assert in the other case that `HasRetBuffer()` is true (since that and `ShouldHaveRetBufArg()` can be out-of-date during import, which I assume it won't be here).